locator/network_topology_strategy.cc

```diff
@@ -325,6 +325,57 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
     co_return tablets;
 }
 
+future<tablet_map> network_topology_strategy::reallocate_tablets_for_existing_table(schema_ptr s, token_metadata_ptr tm) const {
+    // WIP (a copy-paste of the above), basically:
+    // When new RF < old RF, take the most loaded endpoints from load_sketch and drop extra redundant replicas from these endpoints
+    // Otherwise add new replicas from the least loaded endpoints
+    // Pass newReturn new tablet_map
+    tablet_map tablets = tm->tablets().get_tablet_map(s->id());
+    tablet_replica_set replicas = tablets.get_tablet_info(s->id()).replicas;
+    std::unordered_map<basic_sstring<char, uint32_t, 15>, size_t> new_rf{{"dc1", 2}};
+
+    if (new_rf == _dc_rep_factor)
+        co_return tablets;
+
+    load_sketch load(tm);
+    co_await load.populate();
+
+
+    if (new_rf < _dc_rep_factor) {
+
+    }
+
+    // FIXME: Don't use tokens to distribute nodes.
+    // The following reuses the existing token-based algorithm used by NetworkTopologyStrategy.
+    assert(!tm->sorted_tokens().empty());
+    auto token_range = tm->ring_range(dht::token::get_random_token());
+
+    for (tablet_id tb : tablets.tablet_ids()) {
+        natural_endpoints_tracker tracker(*tm, _dc_rep_factor);
+
+        while (true) {
+            co_await coroutine::maybe_yield();
+            if (token_range.begin() == token_range.end()) {
+                token_range = tm->ring_range(dht::minimum_token());
+            }
+            locator::host_id ep = *tm->get_endpoint(*token_range.begin());
+            token_range.drop_front();
+            if (tracker.add_endpoint_and_check_if_done(ep)) {
+                break;
+            }
+        }
+
+        for (auto&& ep : tracker.replicas()) {
+            replicas.emplace_back(tablet_replica{ep, load.next_shard(ep)});
+        }
+
+        tablets.set_tablet(tb, tablet_info{std::move(replicas)});
```

Comment:
Reallocation should preserve existing replicas, and only add new (empty) ones. This overrides all replicas, which may change them to a completely different set.

Reply:
I was thinking about implementing this piece of logic (add or drop new replicas) as something like that:
```
    tablet_replica_set replicas = tablets.get_tablet_info(s->id()).replicas;
    load_sketch load(tm);
    co_await load.populate();
    // add/drop replicas using load_sketch
    tablets.set_tablet(tb, tablet_info{std::move(replicas)});
```
If `tablets.set_tablet(tb, tablet_info{std::move(replicas)});` recreates replicas* from scratch, thus not preserving the replicas we don't want to touch, then which method should I use, given I fetched the old `tablet_replica_set` and added/dropped replicas from it? Should I e.g. use `tablet_map::set_tablet_transition_info` instead? It contains `tablet_replica_set next;` member, something which looks like I can use.

\* what happens with the replicas that have been removed from this map, will they be automatically destroyed?